### PR TITLE
Make matches {CTerminologyCode} work in rules

### PR DESCRIPTION
**Pull request by pieterbos**
the matches operator in the rule was only possible for a single path instead of also a full expression. This adds full `expression matches {something}` support, plus corrects CTerminologyCode validation, plus tests

If the form `/path matches {[ac3]}`, this is added to the assertionResult on the property `pathsConstrainedToValueSets`, so anything rendering a form based on these rules can update selection possibilities for a code for that path, for instance to show a different dropdown-list.

### CODE REVIEW

File: grammars/src/main/antlr/adl_rules.g4 [hunk 1]
```diff
@@ -63,7 +63,7 @@ booleanConstraintExpression
     | equalityExpression;
 
 
-booleanConstraint: adlRulesPath SYM_MATCHES ('{' c_primitive_object '}' | CONTAINED_REGEXP );
+booleanConstraint: equalityExpression SYM_MATCHES ('{' c_primitive_object '}' | CONTAINED_REGEXP );
```
Comment (pieterbos) [inline grammars/src/main/antlr/adl_rules.g4:66]: This should be very backwards compatible: the and/or/xor things above aren't affected, so no change in existing rules.

**Comment by pieterbos** [comment]
possible memory consumption problem noted by Jelte, likely because of the new dummy parent object.. Going to test and fix that first before merging this.

**Comment by pieterbos** [comment]
The cause of this was the archetype in the dummy object not being recognized as the same reference as the archetype used before, resulting in an infinite loop. Implemented a bit of a hack to fix this, but it is now fixed.

File: aom/src/main/java/com/nedap/archie/aom/primitives/CTerminologyCode.java [hunk 2]
```diff
@@ -63,17 +66,17 @@ public boolean isValidValue(TerminologyCode value) {
...
-            } else if (constraint.startsWith("ac")) {
-                if(value.getTerminologyId() != null && value.getTerminologyId().equals(constraint)) {
-                    return true;
-                }
-            }
+        if(value != null && value.getTerminologyId() != null && !value.getTerminologyId().equalsIgnoreCase("local") && !AOMUtils.isValueSetCode(value.getTerminologyId())) {
+            //this is a non-local terminology. If a term binding is there, we may be able to validate, if external, we wil not be able to
+            //TODO: check term bindings as well! However, no idea how to handle the URIs, as the value will not be URI. I think?
+            //so return true for now for non-local terminology values
+            return true;
```
Comment (J3173) [inline aom/src/main/java/com/nedap/archie/aom/primitives/CTerminologyCode.java:73]: I'm not sure if we should return true if we don't support this at this moment. I think it is better to return false for now and only return true if we know the value actually matches the constraint.

File: aom/src/main/java/com/nedap/archie/aom/primitives/CTerminologyCode.java [hunk 3]
```diff
@@ -63,17 +66,17 @@ public boolean isValidValue(TerminologyCode value) {
...
         }
+        List<String> values = this.getValueSetExpanded();
```
Comment (J3173) [inline aom/src/main/java/com/nedap/archie/aom/primitives/CTerminologyCode.java:75]: This doesn't seem to work in Operational Templates when the rules come from an inner archetype. It tries to get the valueset from the wrong archetype.

**Review by J3173 — CHANGES_REQUESTED** [review]

File: aom/src/main/java/com/nedap/archie/aom/primitives/CTerminologyCode.java (hunk 3 above)
Comment (pieterbos) [reply aom/src/main/java/com/nedap/archie/aom/primitives/CTerminologyCode.java:75]: Thanks, that should be fixable, I'll fix that!

File: aom/src/main/java/com/nedap/archie/aom/primitives/CTerminologyCode.java (hunk 2 above)
Comment (pieterbos) [reply aom/src/main/java/com/nedap/archie/aom/primitives/CTerminologyCode.java:73]: that means people cannot use the rm object validator if they rely on external terminology servers or term bindings at all. I consider that to be a problem.

We could make it configurable instead, perhaps?

File: tools/src/test/java/com/nedap/archie/aom/TerminologyCodeConstraintsTest.java [hunk 4]
```diff
@@ -24,15 +51,29 @@ public void noConstraint() {
     @Test
     public void terminologyIdConstraint() {
         CTerminologyCode code = new CTerminologyCode();
+        code.setParent(new DummyRulesPrimitiveObjectParent(archetype));
         code.addConstraint("ac12");
         assertTrue(code.isValidValue(TerminologyCode.createFromString("[ac12::at23]")));
         assertTrue(code.isValidValue(TerminologyCode.createFromString("[ac12::at24]")));
-        assertFalse(code.isValidValue(TerminologyCode.createFromString("[ac13::at23]")));
```
[the remark this answers is not in the document]
Comment (pieterbos) [reply tools/src/test/java/com/nedap/archie/aom/TerminologyCodeConstraintsTest.java:30]: the terminology id in openehr in these cases should be 'local', not 'ac13'. We didn't get that right the first time. This PR corrects that, but still supports validating the old situation. I think we should eventually move all data to 'local' to adhere to standards again.
I could still add a check if it's an ac-code, but know that it won't say much about validity in other implementations.

File: aom/src/main/java/com/nedap/archie/aom/primitives/CTerminologyCode.java (hunk 3 above)
Comment (pieterbos) [reply aom/src/main/java/com/nedap/archie/aom/primitives/CTerminologyCode.java:75]: Turns out, this is pretty hard to fix:
- we flatten the rules into an OPT
- the rules are all at OPT level
- we don't know which rules come from which part of the archetype.
- that makes it hard to determine the correct place to get the ac-code from.

possible solutions: 
- find the nearest for all loop, and use that archetype as context?

Going to try that...

Comment (pieterbos) [reply aom/src/main/java/com/nedap/archie/aom/primitives/CTerminologyCode.java:75]: I did that. it does mean this works:
`/some/path/to/term/code matches {[ac3]}`

But if not a path but an expression that comes from several locations in the archetype, it does not. Not fixable except by doing drastic changes, by adding extra fields to the OPT model.

File: aom/src/main/java/com/nedap/archie/adlparser/treewalkers/RulesParser.java [hunk 5]
```diff
@@ -22,6 +25,10 @@
     public static final Pattern VARIABLE_ASSIGNMENT_PATTERN = Pattern.compile("\\$(?<name>.*)\\:(?<type>.*)");
 
     public RulesParser(ANTLRParserErrors errors) {
+        this(errors, null);
+    }
+
+    public RulesParser(ANTLRParserErrors errors, Archetype archetype) {
```
Comment (J3173) [inline aom/src/main/java/com/nedap/archie/adlparser/treewalkers/RulesParser.java:31]: The archetype isn't used anymore.

File: aom/src/main/java/com/nedap/archie/adlparser/treewalkers/RulesParser.java [hunk 6]
```diff
@@ -181,7 +188,8 @@ private Expression parseBooleanConstraint(BooleanConstraintContext context) {
         } else {
             cPrimitiveObject = primitivesConstraintParser.parseRegex(context.CONTAINED_REGEXP());
         }
-        return new BinaryOperator(ExpressionType.BOOLEAN, OperatorKind.matches, modelReference, new Constraint(cPrimitiveObject));
+        //cPrimitiveObject.setParent(dummyRule); //rules do not yet have a parent, but we'd like to add the archetype
```
Comment (J3173) [inline aom/src/main/java/com/nedap/archie/adlparser/treewalkers/RulesParser.java:191]:
```suggestion
```

File: aom/src/main/java/com/nedap/archie/adlparser/treewalkers/DummyRulesPrimitiveObjectParent.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
+package com.nedap.archie.adlparser.treewalkers;
```
Comment (J3173) [inline aom/src/main/java/com/nedap/archie/adlparser/treewalkers/DummyRulesPrimitiveObjectParent.java:1]: Maybe move this class to the tools module as it is only used there?

File: aom/src/main/java/com/nedap/archie/aom/primitives/CTerminologyCode.java (hunk 2 above)
Comment (J3173) [reply aom/src/main/java/com/nedap/archie/aom/primitives/CTerminologyCode.java:73]: > We could make it configurable instead, perhaps?

That would be nice.

File: tools/src/test/java/com/nedap/archie/rules/evaluation/ParsedRulesEvaluationTest.java [hunk 8]
```diff
@@ -622,5 +653,37 @@ public void stringLiterals() throws Exception {
...
+        Map<String, Object> exampleInstance = generator.generate(opt);
+        Cluster cluster = JacksonUtil.getObjectMapper().readValue(JacksonUtil.getObjectMapper().writeValueAsString(exampleInstance), Cluster.class);
+        //correct case first
+        RuleEvaluation ruleEvaluation = new RuleEvaluation(ArchieRMInfoLookup.getInstance(), JAXBUtil.getArchieJAXBContext(), opt);
+        DvCodedText codedText = (DvCodedText) cluster.itemAtPath("/items[1]/items[1]/value[1]");
+        codedText.setDefiningCode(new CodePhrase(new TerminologyId("local"), "at4"));
+        codedText.setValue("value 1");
+        EvaluationResult result = ruleEvaluation.evaluate(cluster, opt.getRules().getRules());
+        AssertionResult assertionResult = result.getAssertionResults().get(0);
+        assertEquals("ac3", assertionResult.getPathsConstrainedToValueSets().get("/items[id2]/items[id2]/value/defining_code"));
```
Comment (J3173) [inline tools/src/test/java/com/nedap/archie/rules/evaluation/ParsedRulesEvaluationTest.java:675]:
```suggestion
        AssertionResult assertionResult = result.getAssertionResults().get(0);
        assertTrue(assertionResult.getResult());
        assertEquals("ac3", assertionResult.getPathsConstrainedToValueSets().get("/items[id2]/items[id2]/value/defining_code"));
```

You forgot the assert here. It also fails :-(

File: tools/src/test/resources/com/nedap/archie/aom/openEHR-EHR-GENERIC_ENTRY.included.v1.0.0.adls [hunk 9]
```diff
@@ -23,7 +23,13 @@ definition
...
 	        ITEM_TREE[id3] matches {
 	            items matches {
-	                ELEMENT[id2]
+	                ELEMENT[id2] matches {
+	                    value matches {
+	                        DV_CODED_TEXT[id26] matches {
+	                            defining_code matches {[ac3]}
+                            }
+                        }
+	                }
```
Comment (J3173) [inline tools/src/test/resources/com/nedap/archie/aom/openEHR-EHR-GENERIC_ENTRY.included.v1.0.0.adls:32]: The indenting is inconsistent (mixed tabs and spaces).

File: tools/src/test/resources/com/nedap/archie/rules/evaluation/matches_valueset.adls [hunk 10]
```diff
@@ -0,0 +1,70 @@
+archetype (adl_version=2.0.5; rm_release=1.0.2; generated)
```
Comment (J3173) [inline tools/src/test/resources/com/nedap/archie/rules/evaluation/matches_valueset.adls:1]: The indenting in this archetype is inconsistent (mixed tabs and spaces).

File: tools/src/test/resources/com/nedap/archie/rules/evaluation/termcodeparent.adls [hunk 11]
```diff
@@ -0,0 +1,42 @@
+archetype (adl_version=2.0.5; rm_release=1.0.2; generated)
```
Comment (J3173) [inline tools/src/test/resources/com/nedap/archie/rules/evaluation/termcodeparent.adls:1]: The indenting in this archetype is inconsistent (mixed tabs and spaces).

**Review by J3173 — CHANGES_REQUESTED** [review]

**Review by J3173 — CHANGES_REQUESTED** [review]
Nice, one suggestion

File: tools/src/test/java/com/nedap/archie/aom/TerminologyCodeConstraintsTest.java [hunk 12]
```diff
@@ -71,10 +71,14 @@ public void externalTerminology() {
         CTerminologyCode code = new CTerminologyCode();
         code.setParent(new DummyRulesPrimitiveObjectParent(archetype));
         code.addConstraint("ac12");
-        //this is valid, because we have no idea about the binding currently. Could be invalid later, when we have proper
-        //external terminology support
+
+        ValidationConfiguration.setFailOnUnknownTerminologyId(false);
```
Comment (J3173) [inline tools/src/test/java/com/nedap/archie/aom/TerminologyCodeConstraintsTest.java:75]: The import for `ValidationConfiguration` is missing.

**Review by J3173 — CHANGES_REQUESTED** [review]

File: tools/src/test/java/com/nedap/archie/aom/TerminologyCodeConstraintsTest.java (hunk 12 above)
Comment (pieterbos) [reply tools/src/test/java/com/nedap/archie/aom/TerminologyCodeConstraintsTest.java:76]: Sorry about that, fixed!

**Review by J3173 — APPROVED** [review]

**Pull request merged**